src/java/org/apache/cassandra/exceptions/TooManyCachedRowsException.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.exceptions;
+
+import org.apache.cassandra.db.ReadCommand;
+
+import static org.apache.cassandra.exceptions.ExceptionCode.SERVER_ERROR;
+
+/**
+ * Exception thrown when {@link org.apache.cassandra.service.ReplicaFilteringProtection} caches
+ * the configured threshold number of row results from participating replicas.
+ */
+public class TooManyCachedRowsException extends RequestExecutionException
```

Comment:
@adelapena @jrwest This exception is problematic. As a `RequestExecutionException` using the `SERVER_ERROR` code, it isn't logged anywhere. (i.e. It's ignored in `QueryMessage#execute()`.) At the same time, when decoded by tools like `cqlsh`, it presents as a pretty useless `NoHostAvailable`. I'm fishing for options in terms of how to make this work reasonably:

1.) Just leave this as a `RuntimeException`, which makes sure we at least fail and log the exception. The protocol level error ends up being `ServerError`, which doesn't really indicate a read execution problem.

2.) Have it extend `OverloadedException` and log the error explicitly in `ReplicaFilteringProtection`. This would roughly communicate to the client that we breeched some resource usage threshold during a read, but it may not be the intended use for the `OVERLOADED` code?

3.) Leave it as is but make `TooManyCachedRowsException` a first class citizen at the protocol level (as new code and changes to the codec). This would probably still require explicit logging, as w/ #2 above. 

The least painful of the options above seems like #1. Thoughts?

(Note that I never seriously considered `ReadFailureException`, given it has so much extraneous information, and doesn't really indicate an error in coordinator processing.)

Reply:
@jasonstack pointed out to me that #2 actually might make the most sense. `OverloadedException` is actually used to indicate too many in-flight (on-heap) hints and can fail a write operation. This patch is doing something very similar, just for reads.

Reply:
The only problem w/ `OverloadedException` is that tools like `cqlsh` still don't seem to interpret the error usefully. (i.e. Still just dumps a `NoHostAvailable`.)